pipeline-runner/R/gem2s-6-prepare_experiment.R

```diff
@@ -1,221 +1,174 @@
-
-  print_config(6, "Prepare experiment", input, pipeline_config, config)
-
-  # Check which samples have been selected. Otherwiser we are going to use all of them.
-  if (length(config$samples) > 0) {
-    samples <- config$samples
-  } else {
-    samples <- gsub("\\..*", "", list.files("/output/rds_samples"))
-  }
-
-  message("Reloading samples rds for current experiment...")
-  scdata_list <- list()
-  for (sample in samples) {
-    scdata_list[[sample]] <- readRDS(paste("/output/rds_samples/", sample, ".rds", sep = ""))
-  }
-
-  # Merging samples and adding a prefix with the sample name. In pipeline we grep in barcodes to filter by sample.
+#' Prepare experiment for upload to AWS
+#'
+#'  1) Merges the samples for the current experiment
+#'  2) Adds metadata: cellsId, color_pool, and gene annotation
+#'  3) Preparing QC configuration
+#'
+#' @inheritParams download_cellranger
+#' @param prev_out  'output' slot from call to \code{create_seurat}
+#'
+#' @return prev_out \code{prev_out} with added slots 'scdata' containing merged
+#'   \code{SeuratObject} and 'qc_config' containing default config for QC steps.
+#'
+#' @export
+#'
+prepare_experiment <- function(input, pipeline_config, prev_out) {
+  message("Preparing experiment ...")
+
+  scdata_list <- prev_out$scdata_list
+  edrops <- prev_out$edrops
+  samples <- names(scdata_list)
+
+  message("Merging Seurat Objects...")
   if (length(scdata_list) == 1) {
     scdata <- scdata_list[[1]]
-    scdata <- RenameCells(object = scdata, add.cell.id = names(scdata_list)[1])
   } else {
-    scdata <- merge(scdata_list[[1]], y = scdata_list[-1], add.cell.ids = c(samples))
+    scdata <- merge(scdata_list[[1]], y = scdata_list[-1])
   }
 
-  message("Storing gene annotations...")
-  organism <- config$organism
-  annotations <- read.delim("/output/features_annotations.tsv")
+  message("Deduplicating gene annotations...")
+  annot <- prev_out$annot
 
-  # In order to avoid duplicated genes names, we are going to add the ENSEMBL ID for those
-  # genes that are duplicated (geneNameDuplicated-ENSEMBL)
-  gname <- annotations$name
-  # Keep original name in 'original_name' variable
-  annotations$original_name <- gname
+  # add ENSEMBL ID for genes that are duplicated (geneNameDuplicated-ENSEMBL)
+  # original name kept in 'original_name' column
+  gname <- annot$name
+  annot$original_name <- gname
   is.dup <- duplicated(gname) | duplicated(gname, fromLast = TRUE)
-#We need to convert the gene inputs from _ to - bc when we create the Seurat object we do this, and the match would return NA values if any of the inputs still has _. 
-  annotations$input <- gsub('_', '-', annotations$input)
-  annotations$name[is.dup] <- paste(gname[is.dup], annotations$input[is.dup], sep = " - ")
+
+  #We need to convert the gene inputs from _ to - bc when we create the Seurat object we do this, and the match would return NA values if any of the inputs still has _.
+  annot$input <- gsub('_', '-', annot$input)
```

Comment:
why aren't we doing this in the step that is configuring the annotations and then putting then in `prev_out`?

Reply:
Also, why preparing QC configuration here? It doesn't feel like this is the best place for constructing QC config, because it makes that step directly coupled to QC. It will be better to construct QC config just after calling the QC step, or at least, at the bare minimum move this code to a separate file, so that in the future if we want to add another step after "gem2s-prepare-experiment", we don't have to be copy-pasting and moving this QC code around.

Reply:
I would probably move all this config to a file called `qc_config` or something similar. We can have a similar one for all the config that we have for gem2s too.

Reply:
We put it here because QC config is the initial default QC config that QC will be run with.  It gets uploaded to dynamodb and forms the request parameters when QC is initially triggered. We don't have it after calling the QC step because the QC steps don't know if they are running QC for the first time (in which case the parameters are the defaults created here) or the user has changed parameters.

Reply:
I'll move the QC config to another file